h2o-core/src/main/java/hex/AUC2.java

```diff
@@ -263,30 +265,34 @@ public void perRow(double pred, int act, double w ) {
       idx = -idx-1;             // Get index to insert at
 
       // If already full bins, try to instantly merge into an existing bin
-      if( _n > _nBins ) {       // Need to merge to shrink things
+      if (_n == _nBins &&
```

Comment:
This is a bug.  Should merge before it exceeds nBins.  Good catch.

Reply:
Yes, this was the main bug that was causing all the insertions and instantaneous merging of the same bin.

Reply:
Michal:  

The other time consuming work is with reduce and when one try to build a AUC2 object with AUCBuilder.


W

Reply:
Sure, however, this provides 10x speed-up of building the histogram. In my measurements, most of the time is spent in this function (33s) compared to just 1s in reduce().

This only applies to your the original code not to your PR.

This allows to build the histogram with 400 bins for each chunk super fast, it is a good change :) There will be other work needed but that is in your part of the PR.